### PR TITLE
Pubsub tests

### PR DESCRIPTION
**Pull request by jmdobry**
* Refactored Pub/Sub samples into nice command-line programs, like the new Cloud Storage samples
* Brought code Pub/Sub samples code coverage up to 98.66%

### CODE REVIEW

**Comment by jmdobry** [comment]
@jonparrott Can you please review when you have a chance?

File: pubsub/iam.js [hunk 1]
```diff
@@ -13,64 +13,83 @@
...
-var deleteSubscriptionExample = subscriptionSample.deleteSubscriptionExample;
-var pubsub = subscriptionSample.pubsub;
+// [START auth]
+// By default, gcloud will authenticate using the service account file specified
+// by the GOOGLE_APPLICATION_CREDENTIALS environment variable and use the
+// project specified by the GCLOUD_PROJECT environment variable. See
+// https://googlecloudplatform.github.io/gcloud-node/#/docs/guides/authentication
+var gcloud = require('gcloud');
+
+// Get a reference to the pubsub component
```
Comment (jonparrott) [inline pubsub/iam.js]: Should this be `Construct a pubsub client`? or similar?

File: pubsub/iam.js [hunk 2]
```diff
@@ -13,64 +13,83 @@
...
- * @param {string} topicName Name of the topic whose policy is to be retrieved.
- * @param {Function} callback Callback function.
+ * Retrieve a topic's IAM policy.
+ *
+ * @param {string} topicName The name of the topic.
+ * @param {function} callback The callback function.
  */
-function getTopicPolicyExample (topicName, callback) {
+function getTopicPolicy (topicName, callback) {
+  if (!topicName) {
```
Comment (jonparrott) [inline pubsub/iam.js]: Is this check needed?

File: pubsub/iam.js [hunk 3]
```diff
@@ -13,64 +13,83 @@
...
-  topic.iam.getPolicy(function (err, policy, apiResponse) {
+  topic.iam.getPolicy(function (err, policy) {
     if (err) {
       return callback(err);
     }
 
-    // Received the policy
-    console.log(policy); // { etag: 'ACAB' }
-    callback(null, policy, apiResponse);
+    console.log('Got policy:', policy); // { etag: 'ACAB' }
```
Comment (jonparrott) [inline pubsub/iam.js]: You commented about the etag, do you want to give any other details about the returned policy object? A link to the gcloud-nodejs reference would also work.

File: pubsub/iam.js [hunk 4]
```diff
@@ -84,14 +103,14 @@ function setTopicPolicyExample (topicName, callback) {
     ]
   };
 
-  // Retrieve the IAM policy for the provided topic
-  topic.iam.setPolicy(myPolicy, function (err, policy, apiResponse) {
+  // Set the IAM policy for the specified topic
+  topic.iam.setPolicy(myPolicy, function (err, policy) {
```
Comment (jonparrott) [inline pubsub/iam.js]: Nit that can be ignored: I typically always avoid variable names like `mypolicy`. `newPolicy` or just `policy` would work fine.

File: pubsub/iam.js [hunk 5]
```diff
@@ -170,60 +204,80 @@ function testSubscriptionPermissionsExample (subscriptionName, callback) {
...
-    tests,
-    function (err, permissions, apiResponse) {
-      if (err) {
-        return callback(err);
-      }
-
-      console.log('Got permissions for ' + subscriptionName);
-      callback(null, permissions, apiResponse);
+  // Test the IAM policy for the specified subscription
+  subscription.iam.testPermissions(tests, function (err, permissions) {
```
Comment (jonparrott) [inline pubsub/iam.js]: `tests` could be better named. Maybe `permissions`. What's the IAM terminology for those?

File: pubsub/topics.js [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+ */
+function createTopic (topicName, callback) {
+  if (!topicName) {
+    return callback(new Error('"topicName" is required!'));
+  }
+
+  var topic = pubsub.topic(topicName);
+
+  // Get the topic if it exists. Create it if it does not exist.
+  topic.get({
```
Comment (jonparrott) [inline pubsub/topics.js]: Is this the most idiomatic way to do this? Is there not just a 'topic.create()`?

File: pubsub/iam.js (hunk 3 above)
Comment (jmdobry) [inline pubsub/iam.js]: Removed comment, don't want to give details.

File: pubsub/iam.js (hunk 4 above)
Comment (jmdobry) [inline pubsub/iam.js]: Fixed.

File: pubsub/iam.js (hunk 1 above)
Comment (jmdobry) [inline pubsub/iam.js]: Rephrased.

File: pubsub/iam.js (hunk 5 above)
Comment (jmdobry) [inline pubsub/iam.js]: `permissions`, but I will rename to `permissionsToTest` so as not to shadow the `permissions` variable that appears later

File: pubsub/topics.js (hunk 6 above)
Comment (jmdobry) [inline pubsub/topics.js]: This is the only way to do a "get or create"

**Comment by jonparrott** [comment]
LGTM, perhaps have @ace-n take a once-over as well?

File: pubsub/iam.js [hunk 7]
```diff
@@ -13,69 +13,88 @@
...
-  topic.iam.getPolicy(function (err, policy, apiResponse) {
+  topic.iam.getPolicy(function (err, policy) {
     if (err) {
       return callback(err);
     }
 
-    // Received the policy
-    console.log(policy); // { etag: 'ACAB' }
-    callback(null, policy, apiResponse);
+    console.log('Got policy:', policy);
```
Comment (ace-n) [inline pubsub/iam.js]: Should this be "got **topic** policy" to differentiate from the other functions?

File: pubsub/iam.js [hunk 8]
```diff
@@ -13,69 +13,88 @@
...
   // Retrieve the IAM policy for the subscription
-  subscription.iam.getPolicy(function (err, policy, apiResponse) {
+  subscription.iam.getPolicy(function (err, policy) {
     if (err) {
       return callback(err);
     }
 
-    console.log(policy); // { etag: 'ACAB' }
-    callback(null, policy, apiResponse);
+    console.log('Got policy:', policy);
```
Comment (ace-n) [inline pubsub/iam.js]: Ditto on differentiating policies.

File: pubsub/iam.js [hunk 9]
```diff
@@ -84,14 +103,14 @@ function setTopicPolicyExample (topicName, callback) {
...
-  topic.iam.setPolicy(myPolicy, function (err, policy, apiResponse) {
+  // Set the IAM policy for the specified topic
+  topic.iam.setPolicy(newPolicy, function (err, policy) {
     if (err) {
       return callback(err);
     }
 
-    console.log('Updated policy for ' + topicName);
-    callback(null, policy, apiResponse);
+    console.log('Updated policy for: %s', topicName);
```
Comment (ace-n) [inline pubsub/iam.js]: Ditto on differentiating policy types (IAM vs. subscription etc.)

File: pubsub/iam.js [hunk 10]
```diff
@@ -116,114 +139,145 @@ function setSubscriptionPolicyExample (subscriptionName, callback) {
...
-  subscription.iam.setPolicy(myPolicy, function (err, policy, apiResponse) {
+  // Set the IAM policy for the specified subscription
+  subscription.iam.setPolicy(newPolicy, function (err, policy) {
     if (err) {
       return callback(err);
     }
 
-    console.log('Updated policy for ' + subscriptionName);
-    callback(null, policy, apiResponse);
+    console.log('Updated policy for: %s', subscriptionName);
```
Comment (ace-n) [inline pubsub/iam.js]: Ditto on differentiating policy types (IAM vs. subscription etc.)

File: pubsub/iam.js [hunk 11]
```diff
@@ -116,114 +139,145 @@ function setSubscriptionPolicyExample (subscriptionName, callback) {
...
-  topic.iam.testPermissions(tests, function (err, permissions, apiResponse) {
+  // Test the IAM policy for the specified topic
+  topic.iam.testPermissions(permissionsToTest, function (err, permissions) {
     if (err) {
       return callback(err);
     }
 
-    console.log('Got permissions for ' + topicName);
-    callback(null, permissions, apiResponse);
+    console.log('Tested permissions for: %s', topicName);
```
Comment (ace-n) [inline pubsub/iam.js]: Ditto on differentiating policy types (IAM vs. subscription etc.)

File: pubsub/iam.js [hunk 12]
```diff
@@ -116,114 +139,145 @@ function setSubscriptionPolicyExample (subscriptionName, callback) {
...
-      console.log('Got permissions for ' + subscriptionName);
-      callback(null, permissions, apiResponse);
+  // Test the IAM policy for the specified subscription
+  subscription.iam.testPermissions(permissionsToTest, function (err, permissions) {
+    if (err) {
+      return callback(err);
     }
-  );
+
+    console.log('Tested permissions for: %s', subscriptionName);
```
Comment (ace-n) [inline pubsub/iam.js]: Ditto on differentiating policy types (IAM vs. subscription etc.)

File: pubsub/iam.js [hunk 13]
```diff
@@ -116,114 +139,145 @@ function setSubscriptionPolicyExample (subscriptionName, callback) {
...
+  console.log('Usage: node iam RESOURCE COMMAND [ARGS...]');
+  console.log('\nResources:\n');
+  console.log('\ttopics');
+  console.log('\tsubscriptions');
+  console.log('\nCommands:\n');
+  console.log('\tget NAME');
+  console.log('\tset NAME');
+  console.log('\ttest NAME');
+  console.log('\nExamples:\n');
+  console.log('\tnode iam topics get my-topic');
```
Comment (ace-n) [inline pubsub/iam.js]: Should these 6 lines be combined into 2? (by combining the get/set/test variants into one line)

File: pubsub/subscriptions.js [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+
+  var options = {
+    reuseExisting: true
+  };
+  pubsub.subscribe(topicName, subscriptionName, options, function (err, subscription) {
+    if (err) {
+      return callback(err);
+    }
+
+    console.log('Created subscription %s to %s', subscriptionName, topicName);
```
Comment (ace-n) [inline pubsub/subscriptions.js]: At the risk of being pedantic: you might want to say "created subscription %s to **topic** %s" instead.

File: pubsub/system-test/iam.test.js [hunk 15]
```diff
@@ -13,36 +13,75 @@
...
+      });
+    });
+  });
+
+  describe('getSubscriptionPolicy', function () {
+    it('should get a subscriptions\'s policy', function (done) {
+      program.getSubscriptionPolicy(subscriptionName, function (err, policy) {
+        assert.ifError(err);
+        assert(policy);
+        assert(console.log.calledWith('Got policy:', policy));
```
Comment (ace-n) [inline pubsub/system-test/iam.test.js]: Specify policy type

File: pubsub/system-test/iam.test.js [hunk 16]
```diff
@@ -13,36 +13,75 @@
...
+      pubsub.topic(topicName).delete(done);
+    });
+  });
+
+  describe('getTopicPolicy', function () {
+    it('should get a topic\'s policy', function (done) {
+      program.getTopicPolicy(topicName, function (err, policy) {
+        assert.ifError(err);
+        assert(policy);
+        assert(console.log.calledWith('Got policy:', policy));
```
Comment (ace-n) [inline pubsub/system-test/iam.test.js]: Specify policy type

File: pubsub/system-test/iam.test.js [hunk 17]
```diff
@@ -13,36 +13,75 @@
...
+      });
+    });
+  });
+
+  describe('testTopicPermissions', function () {
+    it('should test a topic\'s permissions', function (done) {
+      program.testTopicPermissions(topicName, function (err, permissions) {
+        assert.ifError(err);
+        assert(permissions);
+        assert(console.log.calledWith('Tested permissions for: %s', topicName));
```
Comment (ace-n) [inline pubsub/system-test/iam.test.js]: Specify object type (topic vs subscription)

File: pubsub/system-test/iam.test.js [hunk 18]
```diff
@@ -13,36 +13,75 @@
...
+      });
+    });
+  });
+
+  describe('testSubscriptionPermissions', function () {
+    it('should test a subscriptions\'s permissions', function (done) {
+      program.testSubscriptionPermissions(subscriptionName, function (err, permissions) {
+        assert.ifError(err);
+        assert(permissions);
+        assert(console.log.calledWith('Tested permissions for: %s', subscriptionName));
```
Comment (ace-n) [inline pubsub/system-test/iam.test.js]: Specify object type (topic vs subscription)

File: pubsub/system-test/subscriptions.test.js [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+
+'use strict';
+
+var uuid = require('node-uuid');
+var gcloud = require('gcloud');
+var pubsub = gcloud.pubsub();
+var program = require('../subscriptions');
+var topicName = 'nodejs-docs-samples-test-' + uuid.v4();
+var subscriptionName = 'nodejs-docs-samples-test-sub-' + uuid.v4();
+var projectId = process.env.GCLOUD_PROJECT;
```
Comment (ace-n) [inline pubsub/system-test/subscriptions.test.js]: Do we check that this isn't **null** or unset before using it?

File: pubsub/system-test/subscriptions.test.js [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+          assert(console.log.calledWith('Found %d subscriptions!', allSubscriptions.length));
+          done();
+        });
+      });
+    });
+  });
+
+  describe('pull', function () {
+    before(function (done) {
+      pubsub.topic(topicName).publish({ data: 'Hello World!' }, done);
```
Comment (ace-n) [inline pubsub/system-test/subscriptions.test.js]: Should this use `data: expected` instead?

File: pubsub/test/iam.test.js [hunk 21]
```diff
@@ -14,159 +14,278 @@
 'use strict';
 
 var proxyquire = require('proxyquire').noCallThru();
+var topicName = 'foo';
+var subscriptionName = 'bar';
+var policyMock = 'policy';
+var permissionsMock = 'permissions';
+
+function getSample () {
+  var subscriptionMock = {
```
Comment (ace-n) [inline pubsub/test/iam.test.js]: Any reason why **subscriptionMock** and **topicMock** are the same thing? Is there some way to create a copy or something other than writing the whole thing out again?

File: pubsub/test/iam.test.js [hunk 22]
```diff
@@ -14,159 +14,278 @@
...
-        }
-      }).getSubscriptionPolicyExample('test-subscription', function (err) {
-        assert(err === 'error');
+  describe('getSubscriptionPolicy', function () {
+    it('should get a subscription\'s policy', function () {
+      var sample = getSample();
+      sample.program.getSubscriptionPolicy(subscriptionName, function (err, policy) {
+        assert.ifError(err);
+        assert.equal(policy, policyMock);
+        assert(console.log.calledWith('Got policy:', policyMock));
```
Comment (ace-n) [inline pubsub/test/iam.test.js]: Policy type

File: pubsub/test/iam.test.js [hunk 23]
```diff
@@ -14,159 +14,278 @@
...
-        }
-      }).setSubscriptionPolicyExample('test-subscription', function (err) {
-        assert(err === 'error');
+  describe('setSubscriptionPolicy', function () {
+    it('should set a subscription\'s policy', function () {
+      var sample = getSample();
+      sample.program.setSubscriptionPolicy(subscriptionName, function (err, policy) {
+        assert.ifError(err);
+        assert.equal(policy, policyMock);
+        assert(console.log.calledWith('Updated policy for: %s', subscriptionName));
```
Comment (ace-n) [inline pubsub/test/iam.test.js]: Policy type

File: pubsub/test/iam.test.js [hunk 24]
```diff
@@ -14,159 +14,278 @@
...
-        }
-      }).testTopicPermissionsExample('test-topic', function (err) {
-        assert(err === 'error');
+  describe('testTopicPermissions', function () {
+    it('should test a topic\'s permissions', function () {
+      var sample = getSample();
+      sample.program.testTopicPermissions(topicName, function (err, permissions) {
+        assert.ifError(err);
+        assert.equal(permissions, permissionsMock);
+        assert(console.log.calledWith('Tested permissions for: %s', topicName));
```
Comment (ace-n) [inline pubsub/test/iam.test.js]: Policy type

File: pubsub/test/iam.test.js [hunk 25]
```diff
@@ -14,159 +14,278 @@
...
+      assert(console.log.calledWith('Usage: node iam RESOURCE COMMAND [ARGS...]'));
+      assert(console.log.calledWith('\nResources:\n'));
+      assert(console.log.calledWith('\ttopics'));
+      assert(console.log.calledWith('\tsubscriptions'));
+      assert(console.log.calledWith('\nCommands:\n'));
+      assert(console.log.calledWith('\tget NAME'));
+      assert(console.log.calledWith('\tset NAME'));
+      assert(console.log.calledWith('\ttest NAME'));
+      assert(console.log.calledWith('\nExamples:\n'));
+      assert(console.log.calledWith('\tnode iam topics get my-topic'));
```
Comment (ace-n) [inline pubsub/test/iam.test.js]: I don't suppose it's worth combining some of these lines? e.g. `node iam topics [get/set/test] my-topic`

File: pubsub/test/iam.test.js [hunk 26]
```diff
@@ -14,159 +14,278 @@
...
+
+      sinon.stub(program, 'testSubscriptionPermissions');
+      program.main(['subscriptions', 'test']);
+      assert(program.testSubscriptionPermissions.calledOnce);
+
+      sinon.stub(program, 'printUsage');
+      program.main(['--help']);
+      assert(program.printUsage.calledOnce);
+
+      program.main(['topics', '--help']);
```
Comment (ace-n) [inline pubsub/test/iam.test.js]: Should we check that testTopicPermissions() wasn't called as well here, or is that overkill?

File: pubsub/test/subscriptions.test.js [hunk 27]
```diff
@@ -0,0 +1,246 @@
...
+    it('should print usage', function () {
+      var program = getSample().program;
+
+      program.printUsage();
+
+      assert(console.log.calledWith('Usage: node subscriptions COMMAND [ARGS...]'));
+      assert(console.log.calledWith('\nCommands:\n'));
+      assert(console.log.calledWith('\tcreate TOPIC_NAME SUBSCRIPTION_NAME'));
+      assert(console.log.calledWith('\tdelete SUBSCRIPTION_NAME'));
+      assert(console.log.calledWith('\tpull SUBSCRIPTION_NAME'));
```
Comment (ace-n) [inline pubsub/test/subscriptions.test.js]: Should we combine some of these commands (for brevity), or leave them be?

File: pubsub/test/topics.test.js [hunk 28]
```diff
@@ -0,0 +1,219 @@
...
+    it('should print usage', function () {
+      var program = getSample().program;
+
+      program.printUsage();
+
+      assert(console.log.calledWith('Usage: node topics COMMAND [ARGS...]'));
+      assert(console.log.calledWith('\nCommands:\n'));
+      assert(console.log.calledWith('\tcreate TOPIC_NAME'));
+      assert(console.log.calledWith('\tdelete TOPIC_NAME'));
+      assert(console.log.calledWith('\tpublish TOPIC_NAME MESSAGE'));
```
Comment (ace-n) [inline pubsub/test/topics.test.js]: Should we combine some of these commands?

File: pubsub/topics.js [hunk 29]
```diff
@@ -0,0 +1,181 @@
...
+}
+// [END list_topics]
+
+// [START usage]
+function printUsage () {
+  console.log('Usage: node topics COMMAND [ARGS...]');
+  console.log('\nCommands:\n');
+  console.log('\tcreate TOPIC_NAME');
+  console.log('\tdelete TOPIC_NAME');
+  console.log('\tpublish TOPIC_NAME MESSAGE');
```
Comment (ace-n) [inline pubsub/topics.js]: Should we combine these?

File: pubsub/iam.js (hunk 13 above)
Comment (jmdobry) [inline pubsub/iam.js]: For the user experience I prefer explicit and unambiguous

File: pubsub/topics.js (hunk 29 above)
Comment (jmdobry) [inline pubsub/topics.js]: No

File: pubsub/test/iam.test.js (hunk 26 above)
Comment (jmdobry) [inline pubsub/test/iam.test.js]: Overkill

File: pubsub/iam.js (hunk 7 above)
Comment (jmdobry) [inline pubsub/iam.js]: I will differentiate

File: pubsub/subscriptions.js (hunk 14 above)
Comment (jmdobry) [inline pubsub/subscriptions.js]: sure

File: pubsub/system-test/subscriptions.test.js (hunk 19 above)
Comment (jmdobry) [inline pubsub/system-test/subscriptions.test.js]: Yes, in `_setup.js`. It must be set.

File: pubsub/test/iam.test.js (hunk 21 above)
Comment (jmdobry) [inline pubsub/test/iam.test.js]: The two resources have the same methods. Using the same mock for two different resources would introduce additional cognitive overhead for any programmer who needs to read or update this test.

**Comment by ace-n** [comment]
LTGM, assuming my comments have all been addressed.

**Pull request merged**